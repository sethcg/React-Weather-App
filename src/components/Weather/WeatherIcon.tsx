'use client'

import { FunctionComponent, ElementType } from 'react'
import { IconType } from 'react-icons/lib'

import * as Icons from 'react-icons/wi'

type WeatherType = { DayIcon: string; NightIcon: string }

export const WeatherCode: { [id: string]: WeatherType } = {
  // MORE INFORMATION HERE: https://openweathermap.org/weather-conditions
  // Day:     Night:      Description:
  // 01d 	    01n         Clear Skies
  // 02d 	    02n         Few Clouds
  // 03d 	    03n         Scattered Clouds
  // 04d 	    04n         Broken Clouds
  // 09d      09n         Shower Rain
  // 10d 	    10n         Rain
  // 11d 	    11n	        Thunderstorm
  // 13d 	    13n	        Snow
  // 50d 	    50n	        Mist
  '01': { DayIcon: 'WiDaySunny', NightIcon: 'WiMoonWaxingCrescent4' },
  '02': { DayIcon: 'WiCloud', NightIcon: 'WiCloud' },
  '03': { DayIcon: 'WiCloud', NightIcon: 'WiCloud' },
  '04': { DayIcon: 'WiCloudy', NightIcon: 'WiCloudy' },
  '09': { DayIcon: 'WiShowers', NightIcon: 'WiShowers' },
  '10': { DayIcon: 'WiRain', NightIcon: 'WiRain' },
  '11': { DayIcon: 'WiThunderstorm', NightIcon: 'WiThunderstorm' },
  '13': { DayIcon: 'WiSnow', NightIcon: 'WiSnow' },
  '50': { DayIcon: 'WiFog', NightIcon: 'WiFog' },
}

function getIconName(iconCode: string): string {
  const isDay = iconCode.slice(iconCode.length - 1).toUpperCase() === 'D'
  const code = iconCode.slice(0, iconCode.length - 1)

  if (!WeatherCode[code]) return 'WiNa'
  return isDay ? WeatherCode[code].DayIcon : WeatherCode[code].NightIcon
}

interface Props {
  iconCode: string | undefined
}

export const WeatherIcon: FunctionComponent<Props> = ({ iconCode }) => {
  if (iconCode) {
    const name = getIconName(iconCode)
    if (name && Icons[name as keyof IconType]) {
      const IconComponent = Icons[name as keyof IconType] as ElementType
      return <IconComponent data-testid={iconCode} size={24 * 10} strokeWidth={0.1} />
    }
  }
  return <Icons.WiNa data-testid={iconCode} size={24 * 10} strokeWidth={0.01} />
}
