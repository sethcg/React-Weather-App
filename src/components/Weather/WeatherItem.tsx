'use client'

import { FunctionComponent } from 'react'

interface Props {
  header: string
  value: string
  isTemperature?: boolean
}

export const WeatherItem: FunctionComponent<Props> = ({ header, value, isTemperature = false }) => {
  return (
    <div className="flex w-full flex-row gap-2 text-2xl">
      <h1 className="">{`${header}:`}</h1>
      <h1 className="flex flex-row gap-1">
        {value}
        <span className="text-sm font-extrabold">{isTemperature && value ? 'Â°F' : ''}</span>
      </h1>
    </div>
  )
}
