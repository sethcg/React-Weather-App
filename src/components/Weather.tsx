'use client';

import WeatherIcon from './WeatherIcon';
import { CurrentWeather } from './WeatherType';

interface Weather {
  weatherRef: CurrentWeather | null;
}

function getLatitude(weather: CurrentWeather | null): string {
  return weather ? weather.coord.lat.toFixed(3) : '';
}

function getLongitude(weather: CurrentWeather | null): string {
  return weather ? weather.coord.lon.toFixed(3) : '';
}

function roundTemperature(num: number | undefined): string {
  return num ? Math.round(num).toString() : '';
}

function formatDescription(description: string | undefined): string {
  // Capitalize the first letter of the description
  return description ? `${description[0].toUpperCase()}${description.slice(1)}` : '';
}

export default function Weather({ weatherRef }: Weather) {
  return (
    <div className="flex w-full flex-col items-center justify-between gap-8 p-2">
      <div className="flex min-h-[168px] items-center">
        <WeatherIcon iconId={weatherRef?.weather[0]?.id} iconCode={weatherRef?.weather[0]?.icon} />
      </div>
      <div className="flex w-full flex-col items-start gap-4 px-8">
        <div>
          <h1>Latitude: {getLatitude(weatherRef)}</h1>
        </div>
        <div>
          <h1>Longitude: {getLongitude(weatherRef)}</h1>
        </div>
        <div>
          <h1>Description: {formatDescription(weatherRef?.weather[0].description)}</h1>
        </div>
        <div>
          <h1>Min Temp: {`${roundTemperature(weatherRef?.main.temp_min)} °F`}</h1>
        </div>
        <div>
          <h1>Max Temp: {`${roundTemperature(weatherRef?.main.temp_max)} °F`}</h1>
        </div>
      </div>
    </div>
  );
}
