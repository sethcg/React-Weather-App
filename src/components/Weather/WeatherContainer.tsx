'use client'

import { FunctionComponent } from 'react'
import { WeatherIcon } from './WeatherIcon'
import { WeatherItem } from './WeatherItem'
import { CurrentWeather } from './WeatherType'

import * as countryJson from '../../app/lib/country.json'
const countries: Array<Country> = countryJson.data

interface Country {
  Name: string
  Code: string
}

const getLatitude = (weather: CurrentWeather | null): string => {
  return weather ? weather.coord.lat.toFixed(3) : ''
}

const getLongitude = (weather: CurrentWeather | null): string => {
  return weather ? weather.coord.lon.toFixed(3) : ''
}

const roundTemperature = (num: number | undefined): string => {
  return num ? Math.round(num).toString() : ''
}

const getDescription = (description: string | undefined): string => {
  // CAPITALIZE THE FIRST LETTER OF THE DESCRIPTION
  return description ? `${description[0].toUpperCase()}${description.slice(1)}` : ''
}

const getCountryName = (weather: CurrentWeather | null): string => {
  if (!weather) return ''
  const name: string | undefined = countries.find((data: Country) => data.Code == weather.sys.country)?.Name
  return name ?? ''
}

const getLocalityName = (weather: CurrentWeather | null): string => {
  return weather ? `${weather.name}` : ''
}

interface Props {
  weatherRef: CurrentWeather | null
}

export const WeatherContainer: FunctionComponent<Props> = ({ weatherRef }) => {
  return (
    <div className="flex w-full flex-grow flex-col items-center gap-2 p-2">
      <div className="flex min-h-[168px] items-center">
        <WeatherIcon iconId={weatherRef?.weather[0]?.id} iconCode={weatherRef?.weather[0]?.icon} />
      </div>
      <div className="flex w-full flex-col items-start gap-4 px-2">
        <WeatherItem header={'Country'} value={getCountryName(weatherRef)} />
        <WeatherItem header={'Locality'} value={getLocalityName(weatherRef)} />
        <WeatherItem header={'Latitude'} value={getLatitude(weatherRef)} />
        <WeatherItem header={'Longitude'} value={getLongitude(weatherRef)} />
        <WeatherItem header={'Description'} value={getDescription(weatherRef?.weather[0].description)} />
        <WeatherItem header={'Min Temperature'} value={roundTemperature(weatherRef?.main.temp_min)} isTemperature={true} />
        <WeatherItem header={'Max Temperature'} value={roundTemperature(weatherRef?.main.temp_max)} isTemperature={true} />
      </div>
    </div>
  )
}
