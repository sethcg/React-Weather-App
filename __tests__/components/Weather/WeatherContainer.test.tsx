import { render, screen, waitFor } from '@testing-library/react'
import { WeatherContainer } from '../../../src/components/Weather/WeatherContainer'

// MOCK THE CHILD COMPONENTS
jest.mock('../../../src/components/Weather/WeatherItem', () => ({
  WeatherItem: jest.fn(({ children }) => <div data-testid="WeatherItem">{children}</div>),
}))

jest.mock('../../../src/components/Weather/WeatherIcon', () => ({
  WeatherIcon: jest.fn(({ children }) => <div data-testid="WeatherIcon">{children}</div>),
}))

describe('WeatherContainer Snapshot', () => {
  it('Renders WeatherContainer unchanged', async () => {
    const { container } = await waitFor(() => render(<WeatherContainer weatherRef={null} />))

    expect(container).toMatchSnapshot()
  })
})

describe('WeatherContainer', () => {
  it('Renders the correct information to the WeatherContainer', async () => {
    render(<WeatherContainer weatherRef={null} />)

    const weatherIcon = screen.queryByTestId('WeatherIcon')
    const weatherItem = screen.queryAllByTestId('WeatherItem')

    expect(weatherIcon).toBeTruthy()
    expect(weatherItem).toBeTruthy()
    expect(weatherItem).toHaveLength(7)
  })
})
