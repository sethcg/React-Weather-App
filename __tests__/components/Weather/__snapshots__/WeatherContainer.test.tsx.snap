// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherContainer Snapshot Renders WeatherContainer unchanged 1`] = `
<div>
  <div
    class="flex w-full flex-grow flex-col items-center gap-2 p-2"
  >
    <div
      class="flex min-h-[168px] items-center"
    >
      <div
        data-testid="WeatherIcon"
      />
    </div>
    <div
      class="flex w-full flex-col items-start gap-4 px-2"
    >
      <div
        data-testid="WeatherItem"
      />
      <div
        data-testid="WeatherItem"
      />
      <div
        data-testid="WeatherItem"
      />
      <div
        data-testid="WeatherItem"
      />
      <div
        data-testid="WeatherItem"
      />
      <div
        data-testid="WeatherItem"
      />
      <div
        data-testid="WeatherItem"
      />
    </div>
  </div>
</div>
`;
